File: tests/drivers/js/cli_help.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { spawnSync } from "node:child_process";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const DEFAULT_REPO_ROOT = path.resolve(HERE, "../../..");
export const DEFAULT_CLI_HELP_REFERENCE_PATH = path.join(
  HERE,
  "fixtures",
  "serened_cli_help.json",
);

const CATEGORY_RULES = [
  ["server/rest_server/database_path_feature.cpp", "server"],
  ["server/rest_server/endpoint_feature.cpp", "server"],
  ["server/general_server/scheduler_feature.cpp", "server"],
  ["server/storage_engine/search_engine.cpp", "server"],
  ["server/general_server/general_server_feature.cpp", "http"],
  ["server/general_server/ssl_server_feature.cpp", "ssl"],
  ["libs/basics/log_flags.cpp", "log"],
];

const CATEGORY_LABELS = {
  server: "Server",
  http: "HTTP",
  ssl: "TLS / SSL",
  log: "Logging",
};

const CATEGORY_ORDER = ["server", "http", "ssl", "log"];

const SOURCE_FILES = CATEGORY_RULES.map(([file]) => file);

function readFile(root, relative) {
  return fs.readFileSync(path.join(root, relative), "utf8");
}

function findBalancedCall(text, marker, startIndex = 0) {
  const markerIndex = text.indexOf(marker, startIndex);
  if (markerIndex < 0) {
    return null;
  }
  const open = text.indexOf("(", markerIndex + marker.length);
  if (open < 0) {
    return null;
  }

  let depth = 0;
  let quote = null;
  let escaped = false;

  for (let i = open; i < text.length; i += 1) {
    const ch = text[i];
    if (quote) {
      if (escaped) {
        escaped = false;
      } else if (ch === "\\") {
        escaped = true;
      } else if (ch === quote) {
        quote = null;
      }
      continue;
    }

    if (ch === '"' || ch === "'") {
      quote = ch;
      continue;
    }
    if (ch === "(") {
      depth += 1;
    } else if (ch === ")") {
      depth -= 1;
      if (depth === 0) {
        return {
          body: text.slice(open + 1, i),
          end: i + 1,
        };
      }
    }
  }

  return null;
}

function splitTopLevelCommas(text) {
  const parts = [];
  let start = 0;
  let depth = 0;
  let angleDepth = 0;
  let quote = null;
  let escaped = false;

  for (let i = 0; i < text.length; i += 1) {
    const ch = text[i];
    if (quote) {
      if (escaped) {
        escaped = false;
      } else if (ch === "\\") {
        escaped = true;
      } else if (ch === quote) {
        quote = null;
      }
      continue;
    }

    if (ch === '"' || ch === "'") {
      quote = ch;
      continue;
    }
    if (ch === "(" || ch === "{" || ch === "[") {
      depth += 1;
      continue;
    }
    if (ch === ")" || ch === "}" || ch === "]") {
      depth -= 1;
      continue;
    }
    if (ch === "<") {
      angleDepth += 1;
      continue;
    }
    if (ch === ">") {
      angleDepth = Math.max(0, angleDepth - 1);
      continue;
    }
    if (ch === "," && depth === 0 && angleDepth === 0) {
      parts.push(text.slice(start, i).trim());
      start = i + 1;
    }
  }

  parts.push(text.slice(start).trim());
  return parts;
}

function decodeCppString(value) {
  return value
    .replace(/\\n/g, "\n")
    .replace(/\\t/g, "\t")
    .replace(/\\"/g, '"')
    .replace(/\\\\/g, "\\");
}

function cppStringLiterals(text) {
  const parts = [];
  const re = /"((?:\\.|[^"\\])*)"/g;
  let match;
  while ((match = re.exec(text))) {
    parts.push(decodeCppString(match[1]));
  }
  return parts;
}

function normalizeType(rawType) {
  return rawType
    .replace(/\s+/g, " ")
    .replace(/^std::/, "")
    .replace(/std::/g, "")
    .trim();
}

function normalizeDefault(rawDefault) {
  const value = rawDefault.trim();
  const strings = cppStringLiterals(value);
  if (strings.length > 0) {
    return strings.join("");
  }
  if (value === "{}") {
    return "[]";
  }
  return value;
}

function normalizeDescription(text) {
  return text.replace(/\s+/g, " ").trim();
}

function categoryForSource(relativePath) {
  for (const [suffix, category] of CATEGORY_RULES) {
    if (relativePath.endsWith(suffix)) {
      return category;
    }
  }
  return "other";
}

function parseAbslFlagsFromSource(repoRoot = DEFAULT_REPO_ROOT) {
  const flags = [];

  for (const relative of SOURCE_FILES) {
    const source = readFile(repoRoot, relative);
    let cursor = 0;
    while (true) {
      const call = findBalancedCall(source, "ABSL_FLAG", cursor);
      if (!call) {
        break;
      }
      cursor = call.end;
      const parts = splitTopLevelCommas(call.body);
      if (parts.length < 4) {
        continue;
      }

      const [rawType, rawName, rawDefault, ...rawDescription] = parts;
      const category = categoryForSource(relative);
      flags.push({
        name: rawName.trim(),
        flag: `--${rawName.trim()}`,
        type: normalizeType(rawType),
        defaultValue: normalizeDefault(rawDefault),
        description: normalizeDescription(
          cppStringLiterals(rawDescription.join(",")).join(" "),
        ),
        source: relative,
        category,
        categoryLabel: CATEGORY_LABELS[category] || category,
      });
    }
  }

  return flags.sort((a, b) => {
    const categoryDiff =
      CATEGORY_ORDER.indexOf(a.category) - CATEGORY_ORDER.indexOf(b.category);
    return categoryDiff || a.name.localeCompare(b.name);
  });
}

function parseListSection(lines, heading) {
  const start = lines.findIndex((line) => line.trim().startsWith(heading));
  if (start < 0) {
    return [];
  }

  const items = [];
  let current = null;
  for (const line of lines.slice(start + 1)) {
    if (!line.trim()) {
      break;
    }
    const match = line.match(/^\s*(--\S+)\s+(.*)$/);
    if (match) {
      current = { flag: match[1], description: match[2].trim() };
      items.push(current);
    } else if (current) {
      current.description = `${current.description} ${line.trim()}`.trim();
    }
  }
  return items;
}

function commandDescription(command) {
  if (command.includes(" shell ")) {
    return "Open the embedded DuckDB shell without starting the SereneDB server.";
  }
  if (command.includes(" psql ")) {
    return "Open the psql-compatible shell subcommand without starting the SereneDB server.";
  }
  return "Start the SereneDB pg-wire server, using the optional data directory and flags.";
}

function parseUsageSections(usageMessage) {
  const lines = usageMessage.split(/\r?\n/);
  const title = lines.find((line) => line.trim())?.trim() || "serened";
  const usageStart = lines.findIndex((line) => line.trim() === "Usage:");
  const usage = [];
  if (usageStart >= 0) {
    for (const line of lines.slice(usageStart + 1)) {
      if (!line.trim()) {
        break;
      }
      usage.push(line.trim());
    }
  }

  return {
    title,
    usage,
    commands: usage.map((command) => ({
      command,
      description: commandDescription(command),
    })),
    flagSources: parseListSection(lines, "Flag sources"),
    helpVariants: parseListSection(lines, "Help variants"),
  };
}

function readUsageMessageFromSource(repoRoot = DEFAULT_REPO_ROOT) {
  const source = readFile(repoRoot, "libs/app/app_server.cpp");
  const call = findBalancedCall(source, "absl::SetProgramUsageMessage");
  if (!call) {
    throw new Error("Could not find absl::SetProgramUsageMessage");
  }
  return cppStringLiterals(call.body).join("");
}

function parseFlagsFromHelpOutput(helpText) {
  const flags = [];
  let currentGroup = "other";
  let current = null;

  for (const line of helpText.split(/\r?\n/)) {
    const groupMatch = line.match(/^Flags from (.*):\s*$/);
    if (groupMatch) {
      currentGroup = groupMatch[1].trim();
      current = null;
      continue;
    }

    const flagMatch = line.match(/^\s+--?([A-Za-z0-9_]+)(?:\s+\((.*))?$/);
    if (flagMatch) {
      current = {
        name: flagMatch[1],
        flag: `--${flagMatch[1]}`,
        type: "",
        defaultValue: "",
        description: (flagMatch[2] || "").replace(/\)?;?$/, "").trim(),
        source: currentGroup,
        category: currentGroup,
        categoryLabel: CATEGORY_LABELS[currentGroup] || currentGroup,
      };
      flags.push(current);
      continue;
    }

    if (!current) {
      continue;
    }

    const defaultMatch = line.match(/^\s+default:\s*(.*)$/);
    if (defaultMatch) {
      current.defaultValue = defaultMatch[1].replace(/;$/, "").trim();
    } else if (line.trim()) {
      current.description = `${current.description} ${line.trim()}`
        .replace(/\s+/g, " ")
        .trim();
    }
  }

  return flags;
}

export function parseSerenedHelpOutput(helpText) {
  const sections = parseUsageSections(helpText);
  return {
    generatedFrom: "serened --help",
    ...sections,
    flagGroups: groupFlags(parseFlagsFromHelpOutput(helpText)),
    rawHelp: helpText,
  };
}

function runHelp(binary) {
  const result = spawnSync(binary, ["--help"], {
    encoding: "utf8",
    maxBuffer: 1024 * 1024 * 8,
  });

  if (result.error) {
    if (result.error.code === "ENOENT") {
      return null;
    }
    throw result.error;
  }

  const output = [result.stdout, result.stderr].filter(Boolean).join("\n");
  if (!output.includes("Usage:")) {
    throw new Error(`Help output from ${binary} did not contain a Usage section`);
  }
  return output;
}

function groupFlags(flags) {
  const groups = new Map();
  for (const flag of flags) {
    const key = flag.category || "other";
    if (!groups.has(key)) {
      groups.set(key, {
        id: key,
        label: flag.categoryLabel || CATEGORY_LABELS[key] || key,
        flags: [],
      });
    }
    groups.get(key).flags.push(flag);
  }

  return [...groups.values()].sort((a, b) => {
    const ai = CATEGORY_ORDER.indexOf(a.id);
    const bi = CATEGORY_ORDER.indexOf(b.id);
    if (ai !== -1 || bi !== -1) {
      return (ai === -1 ? 999 : ai) - (bi === -1 ? 999 : bi);
    }
    return a.label.localeCompare(b.label);
  });
}

export function buildCliHelpReference({
  repoRoot = process.env.SDB_CLI_HELP_SOURCE || DEFAULT_REPO_ROOT,
  binary = process.env.SDB_CLI_HELP_BIN || "serened",
  requireBinary = process.env.SDB_CLI_HELP_REQUIRE_BINARY === "1",
} = {}) {
  const helpText = runHelp(binary);
  if (helpText) {
    const parsed = parseSerenedHelpOutput(helpText);
    if (parsed.flagGroups.some((group) => group.flags.length > 0)) {
      return parsed;
    }
  } else if (requireBinary) {
    throw new Error(
      `Could not run ${binary}. Set SDB_CLI_HELP_BIN to a serened binary.`,
    );
  }

  const usageMessage = readUsageMessageFromSource(repoRoot);
  return {
    generatedFrom: "serened source fallback",
    ...parseUsageSections(usageMessage),
    flagGroups: groupFlags(parseAbslFlagsFromSource(repoRoot)),
    rawHelp: usageMessage,
  };
}

export function writeCliHelpReference(outPath, options = {}) {
  const data = buildCliHelpReference(options);
  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  fs.writeFileSync(outPath, `${JSON.stringify(data, null, 2)}\n`);
  return data;
}

export function readCliHelpReference(
  referencePath = DEFAULT_CLI_HELP_REFERENCE_PATH,
) {
  return JSON.parse(fs.readFileSync(referencePath, "utf8"));
}

function stringifyReference(data) {
  return `${JSON.stringify(data, null, 2)}\n`;
}

function diffLines(expectedLines, actualLines) {
  const rows = expectedLines.length + 1;
  const cols = actualLines.length + 1;
  const lengths = Array.from({ length: rows }, () =>
    new Uint32Array(cols),
  );

  for (let i = expectedLines.length - 1; i >= 0; i -= 1) {
    for (let j = actualLines.length - 1; j >= 0; j -= 1) {
      lengths[i][j] =
        expectedLines[i] === actualLines[j]
          ? lengths[i + 1][j + 1] + 1
          : Math.max(lengths[i + 1][j], lengths[i][j + 1]);
    }
  }

  const ops = [];
  let i = 0;
  let j = 0;
  while (i < expectedLines.length && j < actualLines.length) {
    if (expectedLines[i] === actualLines[j]) {
      ops.push({ type: "equal", line: expectedLines[i] });
      i += 1;
      j += 1;
    } else if (lengths[i + 1][j] >= lengths[i][j + 1]) {
      ops.push({ type: "delete", line: expectedLines[i] });
      i += 1;
    } else {
      ops.push({ type: "insert", line: actualLines[j] });
      j += 1;
    }
  }
  while (i < expectedLines.length) {
    ops.push({ type: "delete", line: expectedLines[i] });
    i += 1;
  }
  while (j < actualLines.length) {
    ops.push({ type: "insert", line: actualLines[j] });
    j += 1;
  }

  return ops;
}

export function formatCliHelpReferenceDiff(expected, actual, maxLines = 260) {
  const expectedLines = stringifyReference(expected).split("\n");
  const actualLines = stringifyReference(actual).split("\n");
  const ops = diffLines(expectedLines, actualLines);
  const output = [
    "--- reference: fixtures/serened_cli_help.json",
    "+++ current: serened --help",
  ];
  let equalRun = [];

  function flushEquals() {
    if (equalRun.length === 0) {
      return;
    }
    if (equalRun.length <= 6) {
      output.push(...equalRun.map((line) => ` ${line}`));
    } else {
      output.push(...equalRun.slice(0, 3).map((line) => ` ${line}`));
      output.push(`... ${equalRun.length - 6} unchanged lines ...`);
      output.push(...equalRun.slice(-3).map((line) => ` ${line}`));
    }
    equalRun = [];
  }

  for (const op of ops) {
    if (op.type === "equal") {
      equalRun.push(op.line);
      continue;
    }

    flushEquals();
    output.push(`${op.type === "delete" ? "-" : "+"}${op.line}`);

    if (output.length >= maxLines) {
      output.push(`... diff truncated after ${maxLines} lines ...`);
      break;
    }
  }

  return output.join("\n");
}

export function compareCliHelpReference({
  referencePath = DEFAULT_CLI_HELP_REFERENCE_PATH,
  actual = buildCliHelpReference(),
} = {}) {
  const expected = readCliHelpReference(referencePath);
  const expectedText = stringifyReference(expected);
  const actualText = stringifyReference(actual);

  if (expectedText === actualText) {
    return { matches: true, expected, actual, diff: "" };
  }

  return {
    matches: false,
    expected,
    actual,
    diff: formatCliHelpReferenceDiff(expected, actual),
  };
}

File: tests/drivers/python/cli_help.py
#!/usr/bin/env python3
"""Generate and verify the serened CLI help reference.

Single source of truth: the `serened --help` banner. This tool never scrapes
C++ source -- whatever the binary prints is the contract.

Two modes:

  override   Run `serened --help`, parse it, write the reference JSON.
  check      Run `serened --help`, parse it, diff against the committed
             reference JSON. Exits non-zero on any drift.

The reference JSON is consumed by the docs site (serenedb-site), so its path
and shape are a published contract -- keep them stable.

Usage:
  python3 cli_help.py override [--bin PATH] [--out PATH]
  python3 cli_help.py check    [--bin PATH] [--reference PATH]
"""

from __future__ import annotations

import argparse
import difflib
import json
import os
import re
import subprocess
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
REPO_ROOT = HERE.parents[2]

# The docs site reads this exact path (raw GitHub + local dev). Don't move it
# without updating serene-cli-help-docs in serenedb-site.
DEFAULT_REFERENCE_PATH = (
    REPO_ROOT / "tests" / "drivers" / "python" / "fixtures" / "serened_cli_help.json"
)

DEFAULT_BINARY = os.environ.get(
    "SDB_DRV_SERENED_BIN",
    str(REPO_ROOT / "build" / "bin" / "serened"),
)

# `serened --help` already groups flags under "Flags from <id>:" headings. These
# ids ARE the categories; labels and ordering are presentation only.
CATEGORY_LABELS = {
    "server": "Server",
    "http": "HTTP",
    "ssl": "TLS / SSL",
    "log": "Logging",
}
CATEGORY_ORDER = ["server", "http", "ssl", "log"]


# --------------------------------------------------------------------------
# Running the binary
# --------------------------------------------------------------------------


def run_help(binary: str = DEFAULT_BINARY, *, timeout: float = 15.0) -> str | None:
    """Return the `serened --help` banner, or None if the binary is absent."""
    env = {**os.environ, "NO_COLOR": "1", "TERM": "dumb"}
    try:
        result = subprocess.run(
            [binary, "--help"],
            capture_output=True,
            text=True,
            timeout=timeout,
            env=env,
        )
    except FileNotFoundError:
        return None

    output = "\n".join(part for part in (result.stdout, result.stderr) if part)
    if "Usage:" not in output:
        raise RuntimeError(
            f"Help output from {binary} did not contain a Usage section",
        )
    return output


# --------------------------------------------------------------------------
# Parsing
# --------------------------------------------------------------------------

_PROGNAME_PREFIX = re.compile(r"^\S+:\s+")
_GROUP_HEADING = re.compile(r"^\s*Flags from (.+?):\s*$")
_FLAG_START = re.compile(r"^\s+--([A-Za-z0-9_]+)(?=\s|\(|$)")
_LIST_ITEM = re.compile(r"^\s*(--\S+)\s+(.*)$")
# A reconstructed flag entry: --name (description); default: value;
_FLAG_ENTRY = re.compile(
    r"^--(?P<name>[A-Za-z0-9_]+)\s+\((?P<desc>.*)\);\s*default:\s*(?P<default>.*?);?\s*$"
)


def _command_description(command: str) -> str:
    if " shell " in command:
        return "Open the embedded DuckDB shell without starting the SereneDB server."
    if " psql " in command:
        return "Open the psql-compatible shell subcommand without starting the SereneDB server."
    return "Start the SereneDB pg-wire server, using the optional data directory and flags."


def _parse_list_section(lines: list[str], heading: str) -> list[dict]:
    """Parse a `Heading:` block of `--flag   description` items until a blank line."""
    start = next(
        (i for i, line in enumerate(lines) if line.strip().startswith(heading)),
        -1,
    )
    if start < 0:
        return []

    items: list[dict] = []
    current: dict | None = None
    for line in lines[start + 1:]:
        if not line.strip():
            break
        match = _LIST_ITEM.match(line)
        if match:
            current = {"flag": match.group(1), "description": match.group(2).strip()}
            items.append(current)
        elif current is not None:
            current["description"] = f"{current['description']} {line.strip()}".strip()
    return items


def _parse_usage(lines: list[str]) -> list[str]:
    start = next((i for i, line in enumerate(lines) if line.strip() == "Usage:"), -1)
    if start < 0:
        return []
    usage: list[str] = []
    for line in lines[start + 1:]:
        if not line.strip():
            break
        usage.append(line.strip())
    return usage


def _normalize_default(raw: str) -> str:
    """Strip the trailing `;`, then unwrap a surrounding pair of double quotes."""
    value = raw.strip()
    if len(value) >= 2 and value[0] == '"' and value[-1] == '"':
        value = value[1:-1]
    return value


def _parse_flag_groups(lines: list[str]) -> list[dict]:
    """Parse the `Flags from <id>:` sections into ordered groups of flags.

    Flag entries wrap across lines; we stitch each entry back into one logical
    string before matching, so multi-line descriptions and defaults parse cleanly.
    """
    groups: dict[str, dict] = {}
    current_group: str | None = None
    pending: list[str] = []

    def flush() -> None:
        nonlocal pending
        if not pending or current_group is None:
            pending = []
            return
        entry = re.sub(r"\s+", " ", " ".join(pending)).strip()
        match = _FLAG_ENTRY.match(entry)
        pending = []
        if not match:
            return
        name = match.group("name")
        label = CATEGORY_LABELS.get(current_group, current_group)
        groups[current_group]["flags"].append(
            {
                "name": name,
                "flag": f"--{name}",
                # --help never reports a type; the field stays for schema parity.
                "type": "",
                "defaultValue": _normalize_default(match.group("default")),
                "description": match.group("desc").strip(),
                # --help groups by category only; source mirrors the group id.
                "source": current_group,
                "category": current_group,
                "categoryLabel": label,
            }
        )

    for line in lines:
        heading = _GROUP_HEADING.match(line)
        if heading:
            flush()
            current_group = heading.group(1).strip()
            groups.setdefault(
                current_group,
                {
                    "id": current_group,
                    "label": CATEGORY_LABELS.get(current_group, current_group),
                    "flags": [],
                },
            )
            continue

        if current_group is None:
            continue

        if not line.strip():
            flush()
            current_group = None
            continue

        if _FLAG_START.match(line):
            flush()
            pending = [line.strip()]
        elif pending:
            pending.append(line.strip())

    flush()

    def sort_key(group_id: str) -> tuple:
        order = CATEGORY_ORDER.index(group_id) if group_id in CATEGORY_ORDER else 999
        return (order, groups[group_id]["label"])

    return [groups[key] for key in sorted(groups, key=sort_key)]


def parse_help(help_text: str) -> dict:
    """Turn a `serened --help` banner into the reference structure."""
    lines = help_text.splitlines()

    title_line = next((line.strip() for line in lines if line.strip()), "serened")
    # absl prints "<progname>: <usage message>"; drop the progname prefix.
    title = _PROGNAME_PREFIX.sub("", title_line)

    usage = _parse_usage(lines)

    return {
        "generatedFrom": "serened --help",
        "title": title,
        "usage": usage,
        "commands": [
            {"command": command, "description": _command_description(command)}
            for command in usage
        ],
        "flagSources": _parse_list_section(lines, "Flag sources"),
        "helpVariants": _parse_list_section(lines, "Help variants"),
        "flagGroups": _parse_flag_groups(lines),
    }


# --------------------------------------------------------------------------
# Reference IO + comparison
# --------------------------------------------------------------------------


def build_reference(binary: str = DEFAULT_BINARY) -> dict | None:
    """Parse the reference from the binary, or None if the binary is absent."""
    help_text = run_help(binary)
    if help_text is None:
        return None
    return parse_help(help_text)


def _serialize(data: dict) -> str:
    return json.dumps(data, indent=2, ensure_ascii=False) + "\n"


def write_reference(out_path: Path, data: dict) -> None:
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(_serialize(data), encoding="utf8")


def read_reference(reference_path: Path = DEFAULT_REFERENCE_PATH) -> dict:
    return json.loads(Path(reference_path).read_text(encoding="utf8"))


def diff_reference(expected: dict, actual: dict) -> str:
    """Unified diff of the serialized reference (empty when they match)."""
    return "".join(
        difflib.unified_diff(
            _serialize(expected).splitlines(keepends=True),
            _serialize(actual).splitlines(keepends=True),
            fromfile="reference (committed)",
            tofile="actual (serened --help)",
        )
    )


# --------------------------------------------------------------------------
# CLI
# --------------------------------------------------------------------------


def _cmd_override(args: argparse.Namespace) -> int:
    data = build_reference(args.bin)
    if data is None:
        print(
            f"serened binary not found or not executable: {args.bin}\n"
            "Set SDB_DRV_SERENED_BIN or pass --bin.",
            file=sys.stderr,
        )
        return 1
    out = Path(args.out)
    write_reference(out, data)
    flag_count = sum(len(group["flags"]) for group in data["flagGroups"])
    print(
        f"Wrote {out} from {data['generatedFrom']}: "
        f"{len(data['commands'])} commands, {flag_count} flags"
    )
    return 0


def _cmd_check(args: argparse.Namespace) -> int:
    actual = build_reference(args.bin)
    if actual is None:
        print(f"serened binary not found: {args.bin}; skipping check", file=sys.stderr)
        return 0
    expected = read_reference(Path(args.reference))
    diff = diff_reference(expected, actual)
    if not diff:
        print("serened CLI help reference is up to date.")
        return 0
    print("serened CLI help reference is out of date:\n", file=sys.stderr)
    print(diff, file=sys.stderr)
    print(
        "\nIf this change is expected, regenerate with:\n"
        "  python3 cli_help.py override",
        file=sys.stderr,
    )
    return 1


def main(argv: list[str] | None = None) -> int:
    common = argparse.ArgumentParser(add_help=False)
    common.add_argument("--bin", default=DEFAULT_BINARY, help="path to the serened binary")

    parser = argparse.ArgumentParser(description=__doc__)
    sub = parser.add_subparsers(dest="mode", required=True)

    p_override = sub.add_parser(
        "override", parents=[common], help="generate the reference JSON"
    )
    p_override.add_argument("--out", default=str(DEFAULT_REFERENCE_PATH))
    p_override.set_defaults(func=_cmd_override)

    p_check = sub.add_parser(
        "check", parents=[common], help="diff the binary against the reference JSON"
    )
    p_check.add_argument("--reference", default=str(DEFAULT_REFERENCE_PATH))
    p_check.set_defaults(func=_cmd_check)

    args = parser.parse_args(argv)
    return args.func(args)


if __name__ == "__main__":
    raise SystemExit(main())
